File: cpp/breadth-first-search/WallsAndGates.cpp
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int row, col, distance;
    Node(int row, int col, int distance) {
        this->row = row;
        this->col = col;
        this->distance = distance;
    }
};

class Solution {
public:
    void wallsAndGates(vector<vector<int>>& rooms) {
        int directions[][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
        int rows = rooms.size();
        int cols = rooms[0].size();

        queue<Node> q;

        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (rooms[i][j] == 0) {
                    q.push(Node(i, j, 0));
                }
            }
        }

        while (!q.empty()) {
            Node currNode = q.front();
            q.pop();

            for (auto& direction : directions) {
                int newRow = currNode.row + direction[0];
                int newCol = currNode.col + direction[1];

                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols &&
                    rooms[newRow][newCol] == INT_MAX) {
                    rooms[newRow][newCol] = currNode.distance + 1;
                    q.push(Node(newRow, newCol, currNode.distance + 1));
                }
            }
        }
    }
};

File: python/breadth-first-search/WallsAndGates.py
from typing import List
from collections import deque

class Node:
    def __init__(self, row, col, distance):
        self.row = row
        self.col = col
        self.distance = distance

class Solution:
    def wallsAndGates(self, rooms: List[List[int]]) -> None:
        MAX = 2147483647
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        rows = len(rooms)
        cols = len(rooms[0])

        queue = deque()

        for i in range(rows):
            for j in range(cols):
                if rooms[i][j] == 0:
                    queue.append(Node(i, j, 0))

        while queue:
            curr_node = queue.popleft()

            for direction in directions:
                new_row = curr_node.row + direction[0]
                new_col = curr_node.col + direction[1]
                
                if 0 <= new_row < rows and 0 <= new_col < cols and rooms[new_row][new_col] == MAX:
                    rooms[new_row][new_col] = curr_node.distance + 1
                    queue.append(Node(new_row, new_col, curr_node.distance + 1))
